### PR TITLE
[v2.3] Clarify what different branches are for

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

There's been lotsa confusion about `master` vs `release/v3.1` or whatever. This PR does 2 things to address this:
 - Add a section about branch names to the README (see it at https://github.com/emissary-ingress/emissary/blob/e9b8f766083092f669885f85447d92085e75bd67/README.md)
 - Add a CI job that checks that the tag-derived version number, the branch name, and the `releaseNotes.yml` version are all in agreement.
 
One of the error messages it can spit out is
```
echo "You seem to be on the Git branch for v${branch_majorminor}.z, but Git tags indicate that this is work for v${tag_majorminor}.z"
echo "Perhaps you need to go yell at the person who set up the release branch for the last .y bump."
```
It'd be good if that setting-up-the-branch bit was scripted to reduce the likelihood of messing it up. I'll do that when this repatriates to 3.1; with some changes from 2.3 to then, I don't think it's worth writing a version of it for 2.3.

## Testing

I verified that when I try this [against master](https://github.com/emissary-ingress/emissary/pull/4409) it complains that
```
You seem to be on the Git branch for v3.2.z, but Git tags indicate that this is work for v3.0.z
Perhaps you need to go yell at the person who set up the release branch for the last .y bump.
```
Which is correct because branching was screwy for 3.1.0.

## Checklist

 - [x] I made sure to update `CHANGELOG.md`. - no applicable changes

   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations

 - [x] This is unlikely to impact how Ambassador performs at scale. - not a runtime change

   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability

 - [ ] My change is adequately tested.

   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points

 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

 - [ ] The changes in this PR have been reviewed for security concerns and adherence to security best practices.

### CODE REVIEW

File: .github/workflows/check-branch.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+name: Check branch version
+
+# This is a separate workflow from the main build-and-test workflow
+# because we want it to run when a PR's target branch changes, and
+# it'd be prohibitively to re-run the main workflow every time that
```
Comment (AliceProxy) [inline .github/workflows/check-branch.yml:5]: `# it'd be prohibitively to re-run` -> `# prohibitively expensive/?`

File: README.md [hunk 2]
```diff
@@ -36,6 +34,33 @@ Emissary-ingress enables its users to:
 
 See the full list of [features](https://www.getambassador.io/features/) here.
 
+Branches
```
Comment (AliceProxy) [inline README.md:37]: This section is great!

**Review by AliceProxy — DISMISSED** [review]
lgtm other than what I think is a small typo

File: .github/workflows/check-branch.yml (hunk 1 above)
Comment (LukeShu) [reply .github/workflows/check-branch.yml]: Ah, I meant to include "time-consuming"

Comment (LukeShu) [reply .github/workflows/check-branch.yml]: fixed

**Review by AliceProxy — APPROVED** [review]

**Pull request merged**